lib/sizzle.patched.js

```diff
@@ -1549,20 +1553,151 @@ function setFilters() {}
 setFilters.prototype = Expr.filters = Expr.pseudos;
 Expr.setFilters = new setFilters();
 
+/** Patch #6: Sorting tokens */
+
+/**
+ * Sorts the tokens in order to mitigate the performance issues caused by matching slow pseudos first:
+ * https://github.com/AdguardTeam/ExtendedCss/issues/55#issuecomment-364058745
+ * 
+ * @param {*} tokens a list of tokens to sort
+ */
+var TokenSorter = (function() {
+
+	/**
+	 * Splits compound selector into a list of simple selectors
+	 * 
+	 * @param {*} tokens Tokens to split into groups
+	 * @returns an array consisting of token groups (arrays) and relation tokens.
+	 */
+	var splitCompoundSelector = function(tokens) {
+
+		var groups = [];
+		var currentTokensGroup = [];
+		var maxIdx = tokens.length - 1;
+		for (var i = 0; i <= maxIdx; i++) {
+
+			var token = tokens[i];
+			var relative = Sizzle.selectors.relative[token.type];
+			if (relative) {
+				groups.push(currentTokensGroup);
+				groups.push(token);
+				currentTokensGroup = [];
+			} else {
+				currentTokensGroup.push(token);				
+			}
+
+			if (i === maxIdx) {
+				groups.push(currentTokensGroup);
+			}
+		}
+
+		return groups;
+	};
+
+	var TOKEN_TYPES_VALUES = {
+		// nth-child, etc, always go last
+		"CHILD": 100,
+		"ID": 90,
+		"CLASS": 80,
+		"TAG": 70,
+		"ATTR": 70,
+		"PSEUDO": 60
+	};
+
+	/** 
+	 * A function that defines the sort order.
+	 * Returns a value lesser than 0 if "left" is less than "right".
+	 */
+	var compareFunction = function(left, right) {
+		var leftValue = TOKEN_TYPES_VALUES[left.type];
+		var rightValue = TOKEN_TYPES_VALUES[right.type];
+		return leftValue - rightValue;
+	};
+
+	/**
+	 * Does all the sorting work
+	 * 
+	 * @param {*} tokens a list of tokens to sort
+	 * @returns a re-sorted array of tokens
+	 */
+	var sort = function(tokens) {
+
+		if (!tokens || tokens.length == 1) {
+			return tokens;
+		}
+
+		var sortedTokens = [];
+		var groups = splitCompoundSelector(tokens);
+		for (var i = 0; i < groups.length; i++) {
+			var group = groups[i];
+			if (group instanceof Array) {
+				group.sort(compareFunction);
+				sortedTokens = sortedTokens.concat(group);
+			} else {
+				sortedTokens.push(group);
+			}
+		}
+
+		return sortedTokens;
+	};
+
+	// Expose
+	return {
+		sort: sort
+	}
+})();
+
+/**
+ * Patch #6:
+ * Removes trailing spaces from the tokens list
+ * @param {} tokens tokens to post-process
+ */
+function removeTrailingSpaces(tokens) {
```

Comment:
Do we really need to change a behavior of `tokenize` when `tolerant` option is passed? It is not assumed in the parameter name `tolerant` that a tokenized output will somehow be differently processed. IMO it is more clear to remove trailing descendent combinators from the `ExtendedCssParser`.

Reply:
Also, as of discussing about this, we need to deal with a case where there is a trailing combinator that is not a descendant combinator. Currently it is possible that stylesheets like `div > { display:none!important; }` may not throw in `ExtendedCssParser`, we may process trailing other combinators as a whole in there and throw if there is a non-descendant combinator.

Reply:
> IMO it is more clear to remove trailing descendent combinators from the ExtendedCssParser.

Doing it in the `ExtendedCssParser` leads to caching selectors/tokens with trailing spaces and this kinda defeats the purpose - we want them to be cached properly on the first run.

> we need to deal with a case where there is a trailing combinator that is not a descendant combinator

++

Furthermore, we need to deal with invalid selectors in general. An invalid rule in a stylesheet shouldn't break other extcss rules.

TODO: Deal with invalid selectors

Reply:
>  An invalid rule in a stylesheet shouldn't break other extcss rules.
It is a different issue. What I'm pointing out is that `ExtendedCssParser` should throw on invalid stylesheet, and to do so it is more natural to move the `removeTrailingSpaces` logic to `ExtendedCssParser`

Moreover, `tokenize` function should not be opinionated like this, it should just separate the provided selector string into tokens, and shouldn't arbitrarily ignore part of the input.

Reply:
Your point about caching makes sense, and I just realized that the current way prevents caching in general.

__We always need to pass a 'normalized selector' produced by applying `toSelector` function in Sizzle tokens__  in files other than `ExtendedCssParser`. Without this, attempts to use `tokenize` cache is pointless. Ideally, we should be able to use tokens or compiled matcher function in `Sizzle.select`.

Having this change in mind, I suggest that we should minimize the extent of modifying Sizzle and deal with caching issue thereafter.

Reply:
I need to find some time to think about it before doing anything 

TODO: Review the way `tokenize` is used in the `ExtendedCssParser`

Reply:
Reviewed, changed it, see the commit message